src/test/groovy/org/prebid/server/functional/tests/DsaSpec.groovy

```diff
@@ -0,0 +1,293 @@
+        assert bidderResponse.seatbid[0].bid[0].ext?.dsa == null
+
+
+        and: "PBS should not log warning"
+        assert !response.ext.warnings
+        assert !response.ext.errors
+
+        where:
+        dsaRequired                      | bidExt
+        ReqsDsaRequiredType.NOT_REQUIRED | null
+        ReqsDsaRequiredType.SUPPORTED    | new BidExt(dsa: null)
+    }
+
+    def "AMP request should send DSA to bidder and always succeed when bidder returns DSA"() {
+        given: "Default AmpRequest"
+        def ampRequest = AmpRequest.defaultAmpRequest
+        def dsa = RegsDsa.getDefaultRegsDsa(dsaRequired)
+
+        and: "Stored default bid request with DSA"
+        def ampStoredRequest = BidRequest.defaultBidRequest.tap {
+            regs.ext.dsa = dsa
+            setAccountId(ampRequest.account)
+        }
+
+        and: "Save storedRequest into DB"
+        def storedRequest = StoredRequest.getStoredRequest(ampRequest, ampStoredRequest)
+        storedRequestDao.save(storedRequest)
+
+        and: "Default bidder response with DSA"
+        def bidDsa = BidExtDsa.getDefaultBidExtDsa()
+        def bidResponse = BidResponse.getDefaultBidResponse(ampStoredRequest).tap {
+            seatbid[0].bid[0].ext = new BidExt(dsa: bidDsa)
+        }
+
+        and: "Set bidder response"
+        bidder.setResponse(ampStoredRequest.id, bidResponse)
+
+        when: "PBS processes amp request"
+        def response = defaultPbsService.sendAmpRequest(ampRequest)
+
+        then: "Bidder request should contain DSA"
+        def bidderRequests = bidder.getBidderRequest(ampStoredRequest.id)
+        verifyAll {
+            bidderRequests.regs.ext.dsa.required == dsa.required
+            bidderRequests.regs.ext.dsa.dataToPub == dsa.dataToPub
+            bidderRequests.regs.ext.dsa.pubRender == dsa.pubRender
+            bidderRequests.regs.ext.dsa.transparency[0].domain == dsa.transparency[0].domain
+            bidderRequests.regs.ext.dsa.transparency[0].params == dsa.transparency[0].params
+        }
+
+        and: "Bidder response should contain DSA"
+        def bidderResponse = decode(response.ext.debug.httpcalls.get("generic")[0].responseBody, BidResponse)
+        def actualDsa = bidderResponse.seatbid[0].bid[0].ext.dsa
+        verifyAll {
+            actualDsa.transparency[0].domain == bidDsa.transparency[0].domain
+            actualDsa.transparency[0].params == bidDsa.transparency[0].params
+            actualDsa.adrender == bidDsa.adrender
+            actualDsa.behalf == bidDsa.behalf
+            actualDsa.paid == bidDsa.paid
+        }
+
+        and: "PBS should not log warning"
+        assert !response.ext.warnings
+        assert !response.ext.errors
+
+        where:
+        dsaRequired << [ReqsDsaRequiredType.NOT_REQUIRED, ReqsDsaRequiredType.SUPPORTED, ReqsDsaRequiredType.REQUIRED, ReqsDsaRequiredType.REQUIRED_PUBLISHER_ONLINE_PLATFORM]
```

Comment:
can be simplified:
`        dsaRequired << ReqsDsaRequiredType.values()`

Reply:
I've tried this, but it's not working

Reply:
never mind, it works)